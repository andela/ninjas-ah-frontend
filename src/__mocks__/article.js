export default (Object.article = {
  title: 'my title well done',
  description: 'description of the article: well',
  body: 'body of the article: well'
});
