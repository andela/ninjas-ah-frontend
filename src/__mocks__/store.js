import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import initialState from '../store/initialState';
import apiMiddleware from '../middlewares/apiMiddleware';

export { initialState };
export const middlewares = [thunk, apiMiddleware];
export const mockStore = configureMockStore(middlewares);

export default mockStore(initialState);
