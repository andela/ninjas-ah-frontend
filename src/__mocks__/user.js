export default (Object.user = {
  firstName: 'John',
  lastName: 'Smith',
  username: 'josmi',
  email: 'josmi@email.com',
  image: 'image.jpg'
});

export const userToRegister = {
  ...Object.user,
  password: 'Abcd1234!',
  confirmPassword: 'Abcd1234!'
};

export const mismatchedUserPassword = {
  ...Object.user,
  password: '12345678',
  confirmPassword: '1234abcd'
};

export const wrongUserPassword = {
  ...Object.user,
  password: 'Abcd1234'
};

export const resetPassword = {
  password: '12345678',
  confirmPassword: '12345678'
};

export const mismatchedResetPassword = {
  password: '12345678',
  confirmPassword: '1234abcd'
};

export const matchedResetPassword = {
  passwordOne: 'Brazzaville10!',
  passwordTwo: 'Brazzaville10!'
};

export const sendEmail = { email: 'noreplay@gmail.com' };
export const fakeEmail = { email: 'fsdfsdfsdfdsf--gmail.com' };
