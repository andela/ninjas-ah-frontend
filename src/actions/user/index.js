import * as getUser from './getUser';
import signup from './signup';
import socialAuth from './socialAuth';

export { getUser, signup, socialAuth };
