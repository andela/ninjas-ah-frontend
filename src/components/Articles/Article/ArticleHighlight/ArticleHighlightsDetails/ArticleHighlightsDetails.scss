@import '../../../../../assets/css/_colors';
@import '../../../../../assets/css/_radius';
@import '../../../../../assets/css/_shadow';

.ArticleHighlightsDetails {
  .modal-body {
    .delete-highlight {
      margin-top: 1px;
      width: 60px;
    }
    .no-comment {
      color: $grey;
      font-style: italic;
      opacity: 0.4;
    }
  }
}

@media screen and (max-width: 760px) {
  .ArticleHighlightsDetails {
    .modal-body {
      width: 90%;
    }
  }
}
