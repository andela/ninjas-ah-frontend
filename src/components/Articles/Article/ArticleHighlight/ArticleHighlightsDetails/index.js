import ArticleHighlightDetails from './ArticleHighlightsDetails';

export default ArticleHighlightDetails;
