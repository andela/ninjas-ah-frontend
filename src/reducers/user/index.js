import { user as initialState } from '../../store/initialState';
import getUserReducer from './getUserReducer';
import signupReducer from './signupReducer';
import loginReducer from './loginReducer';
import editProfileReducer from './editProfileReducer';
import uploadImageReducer from './uploadImageReducer';
import forgotPasswordReducer from './forgotPasswordReducer';
import updatePasswordReducer from './updatePasswordReducer';

export default (state = initialState, action) => {
  const getUser = getUserReducer(state, action);
  const signup = signupReducer(state, action);
  const login = loginReducer(state, action);
  const editProfile = editProfileReducer(state, action);
  const uploadImage = uploadImageReducer(state, action);
  const forgotPassword = forgotPasswordReducer(state, action);
  const updatePassword = updatePasswordReducer(state, action);
  return (
    getUser
    || signup
    || editProfile
    || uploadImage
    || login
    || forgotPassword
    || updatePassword
    || state
  );
};
