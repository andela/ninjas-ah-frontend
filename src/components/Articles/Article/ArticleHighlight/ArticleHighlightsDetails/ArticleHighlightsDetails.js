import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';
import profileImagePlaceHolder from '../../../../../assets/images/profile_plaholder.png';
import { Button, Alert, Img } from '../../../../common';
import './ArticleHighlightsDetails.scss';
import { clearCreateRateStore } from '../../../../../actions';
import {
  deleteArticleHighlight,
  clearDeleteArticleHighlightStore
} from '../../../../../actions/articles';

export class ArticleHighlightsDetails extends Component {
  state = {
    comment: '',
    errors: {},
    message: '',
    highlight: { commentAuthor: {} },
    highlightDetailsModalStyle: 'none'
  };

  componentDidMount = () => {
    this.attachClickEventToHighlightedTexts();
  };

  componentDidUpdate = () => {
    this.attachClickEventToHighlightedTexts();
  };

  componentWillReceiveProps = (nextProps) => {
    const { errors, message } = nextProps;
    this.setState(prevState => ({ ...prevState, message, errors }));
  };

  componentWillUnmount = () => {
    const { clearCreateRateStore, clearDeleteArticleHighlightStore } = this.props;
    clearCreateRateStore();
    clearDeleteArticleHighlightStore();
  };

  attachClickEventToHighlightedTexts = () => {
    const elements = document.querySelectorAll('[style*="--highlight"]');
    elements.forEach(element => element.removeEventListener('click', this.showHighlightDetails));
    elements.forEach(element => element.addEventListener('click', this.showHighlightDetails));
  };

  hideModal = (modal) => {
    const { clearDeleteArticleHighlightStore } = this.props;
    clearDeleteArticleHighlightStore();
    this.setState(prevState => ({ ...prevState, comment: '', [modal]: 'none' }));
  };

  showModal = (modal) => {
    const { clearDeleteArticleHighlightStore } = this.props;
    clearDeleteArticleHighlightStore();
    this.setState(prevState => ({ ...prevState, [modal]: 'block' }));
  };

  showHighlightDetails = ({ srcElement }) => {
    const { article: { highlights } } = this.props;
    const { parentNode } = srcElement;
    const styleValue = parentNode.getAttribute('style').replace('--highlight-id:', '');
    const highlightId = styleValue.replace(';', '');
    const highlight = highlights.filter(val => val.id === Number.parseInt(highlightId, 10))[0];

    this.setState(prevState => ({
      ...prevState,
      highlight: highlight || { commentAuthor: {} },
      highlightDetailsModalStyle: (highlight && 'block') || 'none'
    }));
  };

  deleteHighlight = () => {
    const { highlight } = this.state;
    const {
      deleteArticleHighlight,
      clearCreateRateStore,
      article: { slug }
    } = this.props;

    clearCreateRateStore();
    deleteArticleHighlight(slug, highlight.id);
  };

  render() {
    const { profile, article, loading } = this.props;
    const { highlight, message, errors, highlightDetailsModalStyle } = this.state;

    return (
      <div className="ArticleHighlightsDetails">
        <div className="small-screen-4">
          {(message || errors.token || errors.authentication || errors.message) && (
            <Alert
              alertType={(message && 'success') || 'danger'}
              message={message || errors.token || errors.authentication || errors.message}
            />
          )}
        </div>
        <div className={`modal ${highlightDetailsModalStyle}`}>
          <div className="modal-content">
            <div className="modal-header left-align">
              <Button
                buttonClass="button close-highlight-details-modal medium-padding yellow radius-5 text-black"
                onClick={() => this.hideModal('highlightDetailsModalStyle')}
              >
                <FontAwesomeIcon icon={faTimes} size="2x" />
              </Button>
            </div>
            <div className="modal-body">
              {profile.id === highlight.userId || profile.id === article.userId ? (
                <Button
                  onClick={this.deleteHighlight}
                  loading={loading}
                  buttonClass="right delete-highlight button medium-padding danger radius-1 text-white"
                >
                  <FontAwesomeIcon icon={faTimes} />
                </Button>
              ) : (
                ''
              )}
              <div className="medium-v-padding center-align medium-text border b-light-grey radius-1">
                {highlight.comment || <span className="no-comment">No comment</span>}
              </div>
              {profile.id !== highlight.userId ? (
                <div className="center-align medium-text medium-padding">
                  <span className="inline-block medium-padding">
                    <Img
                      imgSrc={highlight.commentAuthor.image || profileImagePlaceHolder}
                      width="45px"
                      imgClass="radius-5"
                    />
                  </span>
                  <br />
                  {highlight.commentAuthor.firstName} {highlight.commentAuthor.lastName}
                </div>
              ) : (
                ''
              )}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ArticleHighlightsDetails.propTypes = {
  profile: PropTypes.object,
  article: PropTypes.object,
  highlight: PropTypes.object,
  loading: PropTypes.bool,
  message: PropTypes.string,
  errors: PropTypes.object,
  clearCreateRateStore: PropTypes.func,
  deleteArticleHighlight: PropTypes.func,
  clearDeleteArticleHighlightStore: PropTypes.func
};

const mapStateToProps = ({
  user: { profile },
  articles: { deleteHighlight: { loading, message, errors } }
}) => ({
  profile,
  loading,
  message,
  errors
});

export default connect(
  mapStateToProps,
  { deleteArticleHighlight, clearDeleteArticleHighlightStore, clearCreateRateStore }
)(ArticleHighlightsDetails);
