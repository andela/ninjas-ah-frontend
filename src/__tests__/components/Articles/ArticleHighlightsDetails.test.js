import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { article, newHighlight } from '../../../__mocks__/article';
import user from '../../../__mocks__/user';
import store from '../../../__mocks__/store';
import ArticleHighlightsDetails, { ArticleHighlightsDetails as ArticleHighlightsDetailsComponent } from '../../../components/Articles/Article/ArticleHighlight/ArticleHighlightsDetails/ArticleHighlightsDetails';
import { mount, shallow } from '../../../../config/enzymeConfig';

const state = {
  comment: '',
  errors: {},
  message: '',
  highlight: { commentAuthor: {} },
  highlightDetailsModalStyle: 'none'
};

const props = {
  profile: { ...user },
  article: { ...article, highlights: [newHighlight] },
  match: { params: { slug: 'slug-slug-slug' } },
  deleteArticleHighlight: jest.fn(() => true),
  clearDeleteArticleHighlightStore: jest.fn(() => true),
  clearCreateRateStore: jest.fn(() => true),
  message: '',
  errors: { message: 'Network error' }
};

const clickEvent = { srcElement: { parentNode: { getAttribute: jest.fn(attribute => attribute) } } };

describe('<ArticleHighlightsDetailsComponent />', () => {
  it('should render without crashing', () => {
    const component = mount(<Provider store={store}>
        <MemoryRouter>
          <ArticleHighlightsDetails {...props} />
        </MemoryRouter>
      </Provider>);
    expect(component).toHaveLength(1);
  });

  it('should render without crashing', () => {
    const component = shallow(<ArticleHighlightsDetailsComponent {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should show a model displaying details of a highlighted text', () => {
    const component = shallow(<ArticleHighlightsDetailsComponent {...props} />);
    component.setProps(props);
    component.setState(state);

    const closeHighlightDetailsButton = component.find('[buttonClass*="close-highlight-details-modal"]');

    component.instance().showModal('highlightDetailsModalStyle');
    expect(component.state().highlightDetailsModalStyle).toEqual('block');

    closeHighlightDetailsButton.simulate('click', {});
    expect(component.state().highlightDetailsModalStyle).toEqual('none');
    expect(component).toMatchSnapshot();
  });

  it('should display details of a highlighted text', () => {
    const component = shallow(<ArticleHighlightsDetailsComponent {...props} />);
    const highlightedText = component.instance().showHighlightDetails(clickEvent);
    expect(component).toMatchSnapshot();
  });

  it('should show a button to delete a highlighted text', () => {
    const component = shallow(<ArticleHighlightsDetailsComponent
        {...{
          ...props,
          profile: { ...user, id: 1 },
          article: { ...article, userId: 1 },
          highlight: { ...props.highlight, userId: 1 }
        }}
      />);

    const deleteHighlightDetailsButton = component.find('[buttonClass*="delete-highlight"]');

    deleteHighlightDetailsButton.simulate('click', {});
    expect(component).toMatchSnapshot();
  });

  it('should not show a button to delete a highlighted text', () => {
    const component = shallow(<ArticleHighlightsDetailsComponent
        {...{
          ...props,
          profile: { ...user, id: 11 },
          article: { ...article, userId: 1 },
          highlight: { ...props.highlight, userId: 1 }
        }}
      />);

    expect(component).toMatchSnapshot();
  });
});
