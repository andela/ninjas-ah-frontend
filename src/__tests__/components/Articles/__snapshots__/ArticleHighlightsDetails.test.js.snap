// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleHighlightsDetailsComponent /> should display details of a highlighted text 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightsDetailsComponent /> should not show a button to delete a highlighted text 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightsDetailsComponent /> should render without crashing 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightsDetailsComponent /> should show a button to delete a highlighted text 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightsDetailsComponent /> should show a model displaying details of a highlighted text 1`] = `ShallowWrapper {}`;
